rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/model/PathConfigs.java

```diff
@@ -99,28 +122,54 @@ private PathConfig buildPathConfig(ISourceLocation projectRoot) {
         }
     }
 
-    private static class PathConfigUpdater extends Thread {
+    private static class WatchRegistration {
+        final ISourceLocation file;
+        final Consumer<ISourceLocationChanged> callback;
+
+        WatchRegistration(ISourceLocation file, Consumer<ISourceLocationChanged> callback) {
+            this.file = file;
+            this.callback = callback;
+        }
+    }
+
+    private class PathConfigUpdater extends Thread {
         private final Map<ISourceLocation, PathConfig> currentPathConfigs;
+        private final Map<ISourceLocation, List<WatchRegistration>> projectWatches;
 
         public PathConfigUpdater(Map<ISourceLocation, PathConfig> currentPathConfigs) {
             super("Path Config updater");
             setDaemon(true);
             this.currentPathConfigs = currentPathConfigs;
+            projectWatches = new ConcurrentHashMap<>();
         }
 
         // we detect changes to roots, and keep track of the last changed time
         // the thread will clear them if the time is longer than the timeout
         private final Map<ISourceLocation, Long> changedRoots = new ConcurrentHashMap<>();
+
+        // Watch a single file. We keep track of the watch registrations so we can unwatch
+        // these files when the project is closed.
```

Comment:
nit: if we turn these in `/** ... */` java docs, the comment is also visible in the UI for down-stream users.